dpc/src/state/builder.rs

```diff
@@ -0,0 +1,371 @@
+
+// The snarkVM library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::prelude::*;
+use snarkvm_utilities::ToBytes;
+
+use anyhow::{anyhow, Result};
+use rand::{CryptoRng, Rng};
+use std::sync::Arc;
+
+#[derive(Clone)]
+pub struct StateBuilder<C: Parameters> {
+    /// A list of given inputs for a state transition.
+    inputs: Vec<Input<C>>,
+    /// A list of expected outputs for a state transition.
+    outputs: Vec<Output<C>>,
+    /// A publicly-visible field with data from the state transition.
+    memo: Vec<u8>,
+    /// A list of errors accumulated from calling the builder.
+    errors: Vec<String>,
+}
+
+impl<C: Parameters> StateBuilder<C> {
+    ///
+    /// Initializes a new instance of `StateBuilder`.
+    ///
+    pub fn new() -> Self {
+        Self {
+            inputs: Vec::new(),
+            outputs: Vec::new(),
+            memo: Vec::new(),
+            errors: Vec::new(),
+        }
+    }
+
+    ///
+    /// Adds the given input into the builder.
+    ///
+    pub fn add_input(mut self, input: Input<C>) -> Self {
+        // Ensure there are no outputs assigned yet, as the builder computes joint serial numbers.
+        if !self.outputs.is_empty() {
+            self.errors.push("Builder cannot add new inputs after outputs".into());
+        }
+
+        match self.inputs.len() < C::NUM_INPUT_RECORDS {
+            true => self.inputs.push(input),
+            false => self.errors.push("Builder exceeded maximum inputs".into()),
+        };
+        self
+    }
+
+    ///
+    /// Adds the given inputs into the builder.
+    ///
+    pub fn add_inputs(mut self, inputs: Vec<Input<C>>) -> Self {
+        for input in inputs {
+            self = self.add_input(input);
+        }
+        self
+    }
+
+    ///
+    /// Adds the given output into the builder.
+    ///
+    pub fn add_output(mut self, output: Output<C>) -> Self {
+        match self.outputs.len() < C::NUM_OUTPUT_RECORDS {
+            true => self.outputs.push(output),
+            false => self.errors.push("Builder exceeded maximum outputs".into()),
+        };
+        self
+    }
+
+    ///
+    /// Adds the given outputs into the builder.
+    ///
+    pub fn add_outputs(mut self, outputs: Vec<Output<C>>) -> Self {
+        for output in outputs {
+            self = self.add_output(output);
+        }
+        self
+    }
+
+    ///
+    /// Appends the given data to the memo field in the builder.
+    ///
+    pub fn append_memo(mut self, data: &Vec<u8>) -> Self {
+        // TODO (howardwu): Change this to not be hardcoded to 64.
+        match self.memo.len() < 64 && (self.memo.len() + data.len()) < 64 {
```

Comment:
Is is safe to have them "=="?

Reply:
Good catch the latter case can, the former cannot

Reply:
Fixed